go-controller/pkg/node/udn_isolation.go

```diff
@@ -0,0 +1,459 @@
+	if err = m.runKubeletRestartTracker(ctx); err != nil {
+		return fmt.Errorf("failed to run kubelet restart tracker: %w", err)
+	}
+	return controller.StartWithInitialSync(m.podInitialSync, m.podController)
+}
+
+func (m *UDNHostIsolationManager) Stop() {
+	controller.Stop(m.podController)
+}
+
+// CleanupUDNHostIsolation removes all nftables chains and sets created by UDNHostIsolationManager.
+func CleanupUDNHostIsolation() error {
+	nft, err := nodenft.GetNFTablesHelper()
+	if err != nil {
+		return fmt.Errorf("failed getting nftables helper: %w", err)
+	}
+	tx := nft.NewTransaction()
+	safeDelete(tx, &knftables.Chain{
+		Name: nodenft.UDNIsolationChain,
+	})
+	safeDelete(tx, &knftables.Set{
+		Name: nftablesUDNPodIPsv4,
+		Type: "ipv4_addr",
+	})
+	safeDelete(tx, &knftables.Set{
+		Name: nftablesUDNPodIPsv6,
+		Type: "ipv6_addr",
+	})
+	return nft.Run(context.TODO(), tx)
+}
+
+func (m *UDNHostIsolationManager) setupUDNFromHostIsolation() error {
+	tx := m.nft.NewTransaction()
+	tx.Add(&knftables.Chain{
+		Name:     nodenft.UDNIsolationChain,
+		Comment:  knftables.PtrTo("Host isolation for user defined networks"),
+		Type:     knftables.PtrTo(knftables.FilterType),
+		Hook:     knftables.PtrTo(knftables.OutputHook),
+		Priority: knftables.PtrTo(knftables.FilterPriority),
+	})
+	tx.Flush(&knftables.Chain{
+		Name: nodenft.UDNIsolationChain,
+	})
+	tx.Add(&knftables.Set{
+		Name:    nftablesUDNPodIPsv4,
+		Comment: knftables.PtrTo("default network IPs of pods in user defined networks (IPv4)"),
+		Type:    "ipv4_addr",
+	})
+	tx.Add(&knftables.Set{
+		Name:    nftablesUDNPodIPsv6,
+		Comment: knftables.PtrTo("default network IPs of pods in user defined networks (IPv6)"),
+		Type:    "ipv6_addr",
+	})
+	m.addRules(tx)
+
+	err := m.nft.Run(context.TODO(), tx)
+	if err != nil {
+		return fmt.Errorf("could not setup nftables rules for UDN from host isolation: %v", err)
+	}
+	return nil
+}
+
+func (m *UDNHostIsolationManager) addRules(tx *knftables.Transaction) {
+	if m.ipv4 {
+		tx.Add(&knftables.Rule{
+			Chain: nodenft.UDNIsolationChain,
+			Rule: knftables.Concat(
+				"socket", "cgroupv2", "level 2", m.kubeletCgroupPath,
+				"ip", "daddr", "@", nftablesUDNPodIPsv4, "accept"),
+		})
+		tx.Add(&knftables.Rule{
+			Chain: nodenft.UDNIsolationChain,
+			Rule: knftables.Concat(
+				"ip", "daddr", "@", nftablesUDNPodIPsv4, "drop"),
+		})
+	}
+	if m.ipv6 {
+		tx.Add(&knftables.Rule{
+			Chain: nodenft.UDNIsolationChain,
+			Rule: knftables.Concat(
+				"socket", "cgroupv2", "level 2", m.kubeletCgroupPath,
+				"ip6", "daddr", "@", nftablesUDNPodIPsv6, "accept"),
+		})
+		tx.Add(&knftables.Rule{
+			Chain: nodenft.UDNIsolationChain,
+			Rule: knftables.Concat(
+				"ip6", "daddr", "@", nftablesUDNPodIPsv6, "drop"),
+		})
+	}
+}
+
+func (m *UDNHostIsolationManager) updateKubeletCgroup() error {
+	tx := m.nft.NewTransaction()
+	tx.Flush(&knftables.Chain{
+		Name: nodenft.UDNIsolationChain,
+	})
+	m.addRules(tx)
+
+	err := m.nft.Run(context.TODO(), tx)
```

Comment:
just flying by but can this block forever since TODO has no limit?

Reply:
yeah I think theoretically it can, but in reality it never happens.
cc @danwinship for more thoughts on this